internal/powerplatform/services/data_record/api_data_record.go

```diff
@@ -0,0 +1,432 @@
+
+				columns[fmt.Sprintf("%s@odata.bind", key)] = fmt.Sprintf("%s/api/data/%s/%s(%s)", environmentUrl, constants.DATAVERSE_API_VERSION, entityDefinition.LogicalCollectionName, dataRecordId)
+			}
+		}
+		if nestedMapList, ok := value.([]interface{}); ok {
+			delete(columns, key)
+			relations[key] = nestedMapList
+		}
+	}
+
+	entityDefinition := getEntityDefinition(ctx, client, environmentUrl, tableName)
+
+	method := "POST"
+	apiPath := fmt.Sprintf("/api/data/%s/%s", constants.DATAVERSE_API_VERSION, entityDefinition.LogicalCollectionName)
+
+	if val, ok := columns[entityDefinition.PrimaryIDAttribute]; ok {
+		method = "PATCH"
+		apiPath = fmt.Sprintf("%s(%s)", apiPath, val)
+	} else if recordId != "" {
+		method = "PATCH"
+		apiPath = fmt.Sprintf("%s(%s)", apiPath, recordId)
+	}
+
+	e, _ := url.Parse(environmentUrl)
+	apiUrl := &url.URL{
+		Scheme: e.Scheme,
+		Host:   e.Host,
+		Path:   apiPath,
+	}
+
+	response, err := client.Api.Execute(ctx, method, apiUrl.String(), nil, columns, []int{http.StatusOK, http.StatusNoContent}, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(response.BodyAsBytes) != 0 {
+		json.Unmarshal(response.BodyAsBytes, &result)
+	} else if response.Response.Header.Get("OData-EntityId") != "" {
+		re := regexp.MustCompile(`\(([a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12})\)`)
+		match := re.FindStringSubmatch(response.Response.Header.Get("OData-EntityId"))
+		if len(match) > 1 {
+			result.Id = match[1]
+		} else {
+			return nil, fmt.Errorf("no entity record id returned from the odata-entityid header")
+		}
+	} else {
+		return nil, fmt.Errorf("no entity record id returned from the API")
+	}
+
+	locationHeader := response.GetHeader("Location")
+	locationHeader = strings.TrimPrefix(locationHeader, fmt.Sprintf("%s/api/data/%s/%s(", environmentUrl, constants.DATAVERSE_API_VERSION, entityDefinition.LogicalCollectionName))
+	locationHeader = strings.TrimSuffix(locationHeader, ")")
+
+	result.Id = locationHeader
+
+	for key, value := range relations {
+		if nestedMapList, ok := value.([]interface{}); ok {
+			e, _ := url.Parse(environmentUrl)
+			apiUrl := &url.URL{
+				Scheme: e.Scheme,
+				Host:   e.Host,
+				Path:   fmt.Sprintf("/api/data/%s/%s(%s)/%s/$ref", constants.DATAVERSE_API_VERSION, entityDefinition.LogicalCollectionName, result.Id, key),
+			}
+
+			existingRelationsResponse := RelationApiResponse{}
+
+			apiResponse, _ := client.Api.Execute(ctx, "GET", apiUrl.String(), nil, nil, []int{http.StatusOK, http.StatusNoContent}, nil)
+
+			json.Unmarshal(apiResponse.BodyAsBytes, &existingRelationsResponse)
+
+			var toBeDeleted []RelationApiBody = make([]RelationApiBody, 0)
+
+			for _, existingRelation := range existingRelationsResponse.Value {
+				delete := true
+				for _, nestedItem := range nestedMapList {
+					nestedMap := nestedItem.(map[string]interface{})
+
+					tableLogicalName, ok := nestedMap["table_logical_name"].(string)
+					if !ok {
```

Comment:
this code feels repeated in the method few times, can you refactor that?

Reply:
done 👍